File: Data Structures & Algorithms/products-of-array-discluding-self/submission-3.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        ans = []
        memory = [1] * len(nums)
        sufx = [1] * len(nums)
        for i in range(len(nums)):
            if i == 0:
                memory[i] = 1
            else:
                memory[i] = nums[i-1]*memory[i-1]
        for i in range(len(nums)-1, -1 , -1):
            if i == len(nums)-1:
                sufx[i] = 1
            else:
                sufx[i] = nums[i+1]*sufx[i+1]
        for i in range(len(nums)):
            ans.append(memory[i] * sufx[i])
        return ans
            

File: Data Structures & Algorithms/top-k-elements-in-list/submission-3.rs
use std::collections::HashMap;
impl Solution {
    pub fn top_k_frequent(nums: Vec<i32>, k: i32) -> Vec<i32> {
        let mut memory: HashMap<i32,i32> = HashMap::new();

        for num in nums{
            *memory.entry(num).or_insert(0) += 1;
        }
        let mut count_vec: Vec<(i32,i32)> = memory.into_iter().collect();
        count_vec.sort_by(|a,b| b.1.cmp(&a.1));
        count_vec.into_iter()
            .take(k as usize)
            .map(|(num, _count)| num)
            .collect()
   
    }
}
